### ISSUE TITLE
Some questions about the shape of A,B,C,D

### ISSUE THREAD

**Issue by Aristo23333** [state-spaces/mamba#524]
Dear Authors,
Thanks for your brilliant works! Now I am learning about the detailed change of parameter shape in your code and in your paper. I noticed that the A is (D,N), which  represents a N*N matrix with diagonal structure, right? It seems to represent there are D N*N matrixes for D channels?  The same situation for D. But I get some confused about the B, C which is (B,L,N) because they seems not to contain the information of different channls. Are they all the same for D channels? Thank you!

**Comment by yuese1234** [comment]
I also have the same question about the B and C matrices. Do you have any new ideas to share?

**Comment by albertfgu** [comment]
Yes, we tied $B$ and $C$ across all the channels. In Mamba-2 we called this "multi-value attention" or "multi-expand SSM" head structure.

**Comment by Aristo23333** [comment]
Thank you for your reply! And I have another question that I find A can change its value during training but maintain a diagonal structure. Have ever explore whether the distribution of these diagonal elements change could make a influence on the performance? May in S4? What I concern most is the distribution of the value, not only the structure. Thank you!

**Comment by Hprairie** [comment]
While not on the distribution of trained weights, I would highly recommend this paper on the parameterization of SSMs https://arxiv.org/pdf/2206.11893, which is a successor to this paper https://arxiv.org/pdf/2008.07669 .

However, I will say in the appendix of Mamba they abalate on the parameterization of selective SSMs and note that random is completely fine. 

Another interesting thing not really mention is that they ensure that A is always negative. To see this look at [this line](https://github.com/state-spaces/mamba/blob/62db608da60f6fc790b8ed9f4b3225e95ca15fde/mamba_ssm/modules/mamba_simple.py#L143). This has to do with ensuring that the recurrent dynamics decay inputs through time.

**Comment by Aristo23333** [comment]
> While not on the distribution of trained weights, I would highly recommend this paper on the parameterization of SSMs https://arxiv.org/pdf/2206.11893, which is a successor to this paper https://arxiv.org/pdf/2008.07669 .
> 
> However, I will say in the appendix of Mamba they abalate on the parameterization of selective SSMs and note that random is completely fine.
> 
> Another interesting thing not really mention is that they ensure that A is always negative. To see this look at [this line](https://github.com/state-spaces/mamba/blob/62db608da60f6fc790b8ed9f4b3225e95ca15fde/mamba_ssm/modules/mamba_simple.py#L143). This has to do with ensuring that the recurrent dynamics decay inputs through time.

Thank you so much for your information and reference

**Comment by TalRub104** [comment]
Hi,
When I print the dimensions of A, B, and C, I see that the dimensions of B and C are (batch_size, 1, d_state, seq_length), which makes sense according to the paper. However, the shape of A is (expand_dim * d_model, d_state), which is weird since A should also be input-dependent after discretization.
Can anyone explain why the dimension of A is not (batch_size, expand_dim * d_model, d_state, seq_length)?
<img width="641" alt="Algo2" src="https://github.com/user-attachments/assets/0f14b26f-d7c1-4c1a-91a5-2accf9a6e34c">

**Comment by Hprairie** [comment]
When you pass it to the function undiscretized. Their hardware aware algorithm will discretize in local SRAM on the GPU to save memory and time. 

Remember that the discretization process is exp(delta * A) where A is the same at all time steps but delta is different. Look at the code and you will notice that we pass both delta and A, meaning the CUDA function will do it for you. 

Hope this helps!

**Comment by TalRub104** [comment]
Thanks!
And the discretization of B also happens in the CUDA function, right?
I see that if I want to get A and B after discretization using Python, I can use the code from the mamba_inner_ref function in selective_scan_interface.py.
I used the function's code up to the for loop to create the following function:

`def get_params_after_discretization(delta, A, B, C, delta_bias=None, delta_softplus=False): #The code of this function was taken from 'selective_scan_ref'
    temp_delta = delta.float()
    if delta_bias is not None:
        temp_delta = temp_delta + delta_bias[..., None].float()
    if delta_softplus:
        temp_delta = F.softplus(temp_delta)

    dim = A.shape[0]
    is_variable_B = B.dim() >= 3
    is_variable_C = C.dim() >= 3
    if A.is_complex():
        print(f"DEBUG: get_params_after_discretization: A.is_complex()")
        if is_variable_B:
            temp_B = torch.view_as_complex(rearrange(B.float(), "... (L two) -> ... L two", two=2))
        if is_variable_C:
            temp_C = torch.view_as_complex(rearrange(C.float(), "... (L two) -> ... L two", two=2))
    else:
        temp_B = B.float()
        temp_C = C.float()

    delta_A = torch.exp(torch.einsum('bdl,dn->bdln', temp_delta, A))
    print(f"DEBUG: get_params_after_discretization temp_B.dim()={temp_B.dim()} temp_C.dim()={temp_C.dim()}")
    if not is_variable_B:
        delta_B = torch.einsum('bdl,dn->bdln', temp_delta, temp_B)
    else:
        if temp_B.dim() == 3:
            delta_B = torch.einsum('bdl,bnl->bdln', temp_delta, temp_B)
        else:
            temp_B = repeat(temp_B, "B G N L -> B (G H) N L", H=dim // temp_B.shape[1])
            delta_B = torch.einsum('bdl,bdnl->bdln', temp_delta, temp_B)
    if is_variable_C and temp_C.dim() == 4:
        temp_C = repeat(temp_C, "B G N L -> B (G H) N L", H=dim // temp_C.shape[1])
    print(f"DEBUG: get_params_after_discretization: delta_A.shape={delta_A.shape} delta_B.shape={delta_B.shape} temp_C.shape={temp_C.shape}")
    return delta_A, delta_B, temp_C`
    
    
I am calling this function right after the call to selective_scan_cuda.fwd.

Do you think this is the correct way to get the parameters after discretization without entering the CUDA function?

**Comment by BouriMohammed** [comment]
So the discretization of B is just delta*B ? and not Inverse((delta*A))*(A - Identity)*delta*B ?

**Comment by Hprairie** [comment]
Nope! If you discretize using Eulers method you get $\Delta B$. We can think of Mamba's discretization as a mixup between eulers method and zero order hold discretization. The derivation of ZOH can be found here https://en.wikipedia.org/wiki/Discretization . Eulers is trivial and pretty simple to do.

**Comment by BouriMohammed** [comment]
So if we use the mamba_simple module, and we want recuperate the $\bar{B_t}$, we need just to compute the Euler transformation ? 

> Nope! If you discretize using Eulers method you get Δ B . We can think of Mamba's discretization as a mixup between eulers method and zero order hold discretization. The derivation of ZOH can be found here https://en.wikipedia.org/wiki/Discretization . Eulers is trivial and pretty simple to do.

**Comment by Hprairie** [comment]
I am not sure if I fully understand the question, but there is no transformation. $\bar B_t = \Delta_t B_t$ is what is used for Mamba. Just from a math perspective, we can consider this as using the $\bar B$ term from Euler's discretization. The reasoning for this is that it is simple to compute and doesn't seem to matter that much. Tri Dao talks about this in another issue, but it's just simple and less computationally expensive to do over using the $\bar B$ term from ZOH's discretization. Hope this makes sense

**Comment by BouriMohammed** [comment]
> I am not sure if I fully understand the question, but there is no transformation. B ¯ t = Δ t B t is what is used for Mamba. Just from a math perspective, we can consider this as using the B ¯ term from Euler's discretization. The reasoning for this is that it is simple to compute and doesn't seem to matter that much. Tri Dao talks about this in another issue, but it's just simple and less computationally expensive to do over using the B ¯ term from ZOH's discretization. Hope this makes sense

Yees it makes sense, thank you. In fact, I'm trying to study the robustness of mamba against input with noise and I need the $\bar{A_t}$ and $\bar{B_t}$. So because I can not recuperate them directly so I need to compute them. So this is why I need to know if I must use ZOH or Euler to compute $\bar{B_t}$.

**Comment by Hprairie** [comment]
Ahhh I see, yes use Euler for $\bar B_t$, i.e. $\bar B_t = \Delta_t B_t$, and ZOH for $\bar A_t $, i.e. $\bar A_t = \exp (\Delta_t A)$. Good luck!

**Comment by BouriMohammed** [comment]
> Ahhh I see, yes use Euler for $\bar B_t$, i.e. $\bar B_t = \Delta_t B_t$, and ZOH for $\bar A_t $, i.e. $\bar A_t = \exp (\Delta_t A)$. Good luck!

Thank you for your response!!!

**Comment by vladyorsh** [comment]
Hi @Hprairie,

thanks for clarifying. Have your or any of your colleagues ran experiments on different B discretizations? I see nothing against the thought that A and B can be discretized differently, but the Euler method can shoot in a foot sometimes since it's the first order method.

**Comment by Hprairie** [comment]
I remember asking Tri Dao about this, and I think he said that he tried it and found that it was unnecessary in practice, though I can't remember for sure. Realistically from a computational point of view, we want to do the simplest discretizations. There is this paper [here](https://arxiv.org/abs/2407.14207) from some people that I know who cast it into an online learning form and discretized it differently. I think they have ablations comparing some discretizations, but not many.

**Comment by albertfgu** [comment]
I tried this quite extensively for a while and never found significant differences. To be honest, I wanted the ZOH B discretization to be better empirically, but it didn't seem to matter so we went with the Euler discretization. Several other people discovered this trick concurrently, by the way, including the LRU team (https://arxiv.org/abs/2303.06349)

I still think it is conceivable that there are regimes (for example for extreme dt values) where the B discretization could matter. Would be very curious to see!

**Comment by vladyorsh** [comment]
Hi @Hprairie, hi @albertfgu,

thank you for the very comprehensive answers, I appreciate that!
